### PR TITLE
Fix NVFP4/MX QAT backward dropping the bias gradient

### PR DESCRIPTION
**Pull request by shoemoney**
The custom autograd functions in `torchao/prototype/qat/nvfp4.py` and `torchao/prototype/qat/mx.py` accept `bias` in forward and pass it to the addmm dispatch, but their backward returns `None` in the bias gradient slot. After `loss.backward()`, `linear.bias.grad` is `None`, so the optimizer never updates any bias for the entire QAT run. There is no error; the model just trains with frozen biases and silently degraded accuracy.

The mainline sibling `FakeQuantizedLinear` (`torchao/quantization/qat/linear.py`, `F.linear(x, w, self.bias)`) trains bias normally through autograd, so only the NVFP4 and MX prototype paths are affected. The existing regression tests for these paths use bias=False models (`M`/`Sub` in `test_qat.py`), which is why they cannot catch this.

Fix: record bias presence on `ctx` in forward, and return the bias gradient in backward: `grad_output.sum(0)` for NVFP4 (grad_output is always 2D there, the module flattens 3D inputs before `apply`) and `grad_output_2d.sum(0)` for MX. Return `None` when bias is absent.

Tests: `M`/`Sub` gain an optional `bias` arg (default False, existing callers unchanged). `test_qat_nvfp4_training` is parametrized over `use_bias` and asserts bias gradients flow and are nonzero. `test_mx_fake_quantized_linear_backward` (already parametrized over bias) now asserts `bias.grad` exists and matches the reference `nn.Linear` bias gradient. Full `test_qat.py`: 141 passed, 48 skipped (CUDA-gated), no failures. Verified the gradient mechanism on CPU by calling the autograd functions directly: bias grad was `None` before and matches `grad_output.sum(0)` after; the new test parametrizations themselves are CUDA-gated like their neighbors.

### CODE REVIEW

**Comment by vkuzo** [comment]
@claude review

**Comment by shoemoney** [comment]
The Check PR Labels failure wants a module: label, which I cannot add as an outside contributor. `module: qat` fits this change if a maintainer can apply it. The test-rocm nightly failure is unrelated to this diff (CPU-only gradient path change).

File: torchao/prototype/qat/mx.py [hunk 1]
```diff
@@ -117,6 +117,7 @@ def forward(
 
         ctx.save_for_backward(_input_2d, weight)
         ctx.orig_shape = orig_shape
+        ctx.bias_present = bias is not None
```
Comment (vkuzo) [inline torchao/prototype/qat/mx.py]: instead of creating a new flag we should be able to use `ctx.needs_input_grad[2]`

**Comment by vkuzo** [comment]
thanks for the fix! I commented with a simplification, after that is addressed this should be good to go

File: torchao/prototype/qat/mx.py (hunk 1 above)
Comment (shoemoney) [reply torchao/prototype/qat/mx.py]: Done, both files now use ctx.needs_input_grad[2] and the flag is gone. Behavior verified unchanged on CPU (bias grad matches grad_output.sum(0) through real autograd for MX end to end, and via direct backward for NVFP4); test_qat.py still 141 passed. It also handles the requires_grad=False case the flag missed. Thanks for the label and the review.

**Comment by vkuzo** [comment]
@claude review

**Review by andrewor14 — APPROVED** [review]

**Comment by andrewor14** [comment]
Looks great, thanks for the fix!
